lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php

```diff
@@ -675,6 +676,18 @@ public function get_custom_templates() {
 		return $custom_templates;
 	}
 
+	/**
+	 * Returns the current theme's wanted patterns(slugs) to be
+	 * registered from Pattern Directory.
+	 *
+	 * @return array|null
+	 */
+	public function get_pattern_settings() {
+		if ( isset( $this->theme_json['patterns'] ) && is_array( $this->theme_json['patterns'] ) ) {
+			return $this->theme_json['patterns'];
+		}
+	}
```

Comment:
Instead of copying the entire file to lib/compat/wordpress-6.0, should we instead keep the original file under `lib/compact/wordpress-5.9` but rename the class to match core classname and only declare it if we're on 5.8 (the class doesn't exist) and in `lib/compat/wordpress-6.0` extend that class (like you did above for the rest controller and I believe I did that on another controller).

The advantage of this is that we know exactly what's on 5.9 and what was added on 6.0

Reply:
It seems in many cases including this one, this is quite hard due to private functions and properties. In this case for making it work we need to override the `VALID_TOP_LEVEL_KEYS` const in child class, but this const is referenced inside private functions in the parent class, so it uses the existing old value.

I then tried to move more functions to the new child class but we have so many private functions/properties there used, that we would need to move so much code there, changing the names(prefix `gutenberg_`) in order to be called by the child class without being private, but the content would be the same.

I think we cannot avoid this easily - at least from what I could see, and maybe we should copy the whole class as is here in `6.0` and in followups check what we should change to `protected` instead of `private`. That would help as from that point on, to handle this more gracefully.

What do you think? Am I missing something php related?

Reply:
Seems like the problem here is the change in `VALID_TOP_LEVEL_KEYS` right? Would be good if we could switch at some point to "schema" based validation and avoid the adhoc code. That said, it's subject for another time.

For now, I guess we don't have any other option than copying everything 🤷

Reply:
In this case `VALID_TOP_LEVEL_KEYS` is affected, but in the future we will probably need to make more changes in other private functions. That goes to other classes as well, so that's why I'm suggesting making some things `protected` or `static` for `6.0` and then we could make adjustments easier for future releases.

Reply:
Personally, I think each PR should add its own needed changes, if we can land this without copying everything, I'd do that. If we need to make another change later, we'll need to figure how to do it at that point.

Reply:
What if we add a couple of filters to the JSON parser in WP-Core for 5.9.1?
That would unblock this PR - as well as some others that currently have no way to hook in there

Reply:
>If we need to make another change later, we'll need to figure how to do it at that point.

That's the thing though, the way I see it it as I [explained above](https://github.com/WordPress/gutenberg/pull/38323#discussion_r798421891), we cannot avoid copying the `class`. But if we do not change some `private` things we will end up copying the class forever(when we have changes of course). Do you find something bad with making some things `protected` or are there strong reasons for keeping them `private`?

Reply:
The required changes in core for this to work with a child class that changes `VALID_TOP_LEVEL_KEYS` and just adds 
the new `get_patterns` function would be:

1. Change $theme_json to `protected`, because we need to access this from the child class
2. Update the line that uses `VALID_TOP_LEVEL_KEYS` from `self::VALID_TOP_LEVEL_KEYS` to `static::VALID_TOP_LEVEL_KEYS`,  as `static::` is inheritance-aware.

Reply:
Incorporated in https://github.com/WordPress/gutenberg/pull/38625 the changes needed for this and pushed at https://github.com/WordPress/gutenberg/pull/38323/commits/f6a6dd72dbbe045b092d480ef9502f17df4169cb and https://github.com/WordPress/gutenberg/pull/38323/commits/52ca23f4d1aa7d672e674ad188b17e6977edccc2